util/src/keys/store.rs

```diff
@@ -0,0 +1,297 @@
+	fn remove<Value: FromRawBytes + BytesConvertable> (&mut self, key: &Key, password: Option<&str>) -> Option<Value>;
+	/// Deletes key-value by key and returns if the key-value existed
+	fn delete(&mut self, key: &Key) -> bool {
+		self.remove::<Bytes>(key, None).is_some()
+	}
+}
+
+/// Error retrieving value from encrypted hashmap
+#[derive(Debug)]
+pub enum EncryptedHashMapError {
+	/// Encryption failed
+	InvalidPassword,
+	/// No key in the hashmap
+	UnknownIdentifier,
+	/// Stored value is not well formed for the requested type
+	InvalidValueFormat(FromBytesError),
+}
+
+/// Represent service for storing encrypted arbitrary data
+pub struct SecretStore {
+	directory: KeyDirectory
+}
+
+impl SecretStore {
+	/// new instance of Secret Store
+	pub fn new() -> SecretStore {
+		let mut path = ::std::env::home_dir().expect("Failed to get home dir");
+		path.push(".keys");
+		SecretStore {
+			directory: KeyDirectory::new(&path)
+		}
+	}
+
+	#[cfg(test)]
+	fn new_test(path: &::tests::helpers::RandomTempPath) -> SecretStore {
+		SecretStore {
+			directory: KeyDirectory::new(path.as_path())
+		}
+	}
+}
+
+fn derive_key_iterations(password: &str, salt: &H256, c: u32) -> (Bytes, Bytes) {
+	let mut h_mac = Hmac::new(::rcrypto::sha2::Sha256::new(), password.as_bytes());
+	let mut derived_key = vec![0u8; KEY_LENGTH_USIZE];
+	pbkdf2(&mut h_mac, &salt.as_slice(), c, &mut derived_key);
+	let derived_right_bits = &derived_key[0..KEY_LENGTH_AES_USIZE];
+	let derived_left_bits = &derived_key[KEY_LENGTH_AES_USIZE..KEY_LENGTH_USIZE];
+	(derived_right_bits.to_vec(), derived_left_bits.to_vec())
+}
+
+fn derive_key(password: &str, salt: &H256) -> (Bytes, Bytes) {
+	derive_key_iterations(password, salt, KEY_ITERATIONS)
+}
+
+fn derive_mac(derived_left_bits: &[u8], cipher_text: &[u8]) -> Bytes {
+	let mut mac = vec![0u8; KEY_LENGTH_AES_USIZE + cipher_text.len()];
+	mac[0..KEY_LENGTH_AES_USIZE].clone_from_slice(derived_left_bits);
+	mac[KEY_LENGTH_AES_USIZE..cipher_text.len()+KEY_LENGTH_AES_USIZE].clone_from_slice(cipher_text);
+	mac
+}
+
+impl EncryptedHashMap<H128> for SecretStore {
+	fn get<Value: FromRawBytes + BytesConvertable>(&self, key: &H128, password: &str) -> Result<Value, EncryptedHashMapError> {
+		match self.directory.get(key) {
+			Some(key_file) => {
+				let decrypted_bytes = match key_file.crypto.kdf {
+					KeyFileKdf::Pbkdf2(ref params) => {
+						let (derived_left_bits, derived_right_bits) = derive_key_iterations(password, &params.salt, params.c);
+						if derive_mac(&derived_right_bits, &key_file.crypto.cipher_text)
+							.sha3() != key_file.crypto.mac { return Err(EncryptedHashMapError::InvalidPassword); }
+
+						let mut val = vec![0u8; key_file.crypto.cipher_text.len()];
+						match key_file.crypto.cipher_type {
+							CryptoCipherType::Aes128Ctr(ref iv) => {
+								crypto::aes::decrypt(&derived_left_bits, &iv.as_slice(), &key_file.crypto.cipher_text, &mut val);
+							}
+						}
+						val
+					}
+					_ => { unimplemented!(); }
```

Comment:
yep
will implement of importing from geth/eth formats and old versions
next part!